### PR TITLE
declaration of products with sacado

### PR DESCRIPTION
**Pull request by asartori86**

### CODE REVIEW

File: include/deal.II/base/template_constraints.h [hunk 1]
```diff
@@ -461,6 +464,63 @@ struct ProductType<unsigned int,float>
   typedef float type;
 };
 
+/**
+ * declaration of products for sacado types
+ */
+#ifdef DEAL_II_WITH_TRILINOS
+typedef Sacado::Fad::DFad<double> fad_double;
+typedef Sacado::Fad::DFad<fad_double> fad_fad_double;
```
Comment (bangerth) [inline include/deal.II/base/template_constraints.h]: Let's not create these typedefs but instead use the fully qualified names below.

**Comment by bangerth** [comment]
This is a reasonable change, but I'm concerned that the inclusion of `sacado.hpp` everywhere will lead to a significant slowdown in compile speed. Can you check how long it takes to compile the library right now, vs with your change?

If it does lead to a significant change in time, I have ideas how we can restructure your code in a way that doesn't impact everything else.

**Comment by asartori86** [comment]
Before that I melt my cpu with two compilations :) coul you share your ideas?

**Comment by bangerth** [comment]
I'm thinking of something like a header file `include/deal.II/base/trilinos/sacado.h` (or similarly named) that `#include`s `sacado.hpp` and declares these class specializations. If you wanted to use Sacado, you'd then include our own header, rather than the Trilinos header right away.

**Comment by asartori86** [comment]
@bangerth, I vote for your idea and no compilations :)

Let me know what to do

**Comment by bangerth** [comment]
In that case go right ahead with my idea ;-)

**Comment by tamiko** [comment]
On Thu, Sep 17, 2015, at 10:47 CDT, Wolfgang Bangerth <notifications@github.com> wrote:

> I'm thinking of something like a header file
> `include/deal.II/base/trilinos/sacado.h` (or similarly named) that
> `#include`s `sacado.hpp` and declares these class specializations. If
> you wanted to use Sacado, you'd then include our own header, rather
> than the Trilinos header right away.

Yes, this is exactly what I wanted to suggest.

**Comment by tamiko** [comment]
@asartori86 
Can you please test the following three instantiations:
```
DEAL_II_NAMESPACE_OPEN                                                             
                                                                                   
template <typename T>                                                              
struct ProductType<Sacado::Fad::DFad<T>, double>                                   
{                                                                                  
  typedef Sacado::Fad::DFad<T> type;                                               
};                                                                                 
                                                                                   
template <typename T>                                                              
struct ProductType<double, Sacado::Fad::DFad<T> >                                  
{                                                                                  
  typedef Sacado::Fad::DFad<T> type;                                               
};                                                                                 
                                                                                   
template <typename T1, typename T2>                                                
struct ProductType<Sacado::Fad::DFad<T1>, Sacado::Fad::DFad<T2> >                  
{                                                                                  
  typedef Sacado::Fad::DFad<typename ProductType<T1, T2>::type> type;              
};                                                                                 
                                                                                   
DEAL_II_NAMESPACE_CLOSE  
```
instead of your manual recursion fad_double, fad_fad_double, fad_fad_fad_double - this should just instantiate all of them.

**Comment by tamiko** [comment]
@bangerth An alternative might be to just forward declare ```Sacado::Fad::DFad```. But of course, this might lead to a maintenance burden with changes for every Trilinos version...

**Comment by bangerth** [comment]
> @bangerth <https://github.com/bangerth> An alternative might be to just
> forward declare |Sacado::Fad::DFad|. But of course, this might lead to a
> maintenance burden with changes for every Trilinos version...

Yea, we don't know (and shouldn't know) the exact declaration of `DFad`.

**Comment by asartori86** [comment]
@bangerth @tamiko : Thanks for your suggestions. I've implemented them in a file named `deal.II/base/sacado_product_type.h`. I've also added three tests and rebased on origin/master

File: include/deal.II/base/sacado_product_type.h [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+#ifdef DEAL_II_WITH_TRILINOS
+#include "Sacado.hpp"
+
+DEAL_II_NAMESPACE_OPEN
+
+template <typename T>
+struct ProductType<Sacado::Fad::DFad<T>, float>
+{
+  typedef Sacado::Fad::DFad<T> type;
```
Comment (bangerth) [inline include/deal.II/base/sacado_product_type.h]: What actually happens when you multiply a `DFad<double>` with, say, a `long double`? Do you get a `DFad<double>`?

File: include/deal.II/base/sacado_product_type.h [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+template <typename T>
+struct ProductType<int, Sacado::Fad::DFad<T> >
+{
+  typedef Sacado::Fad::DFad<T> type;
+};
+
+template <typename T, typename U>
+struct ProductType<Sacado::Fad::DFad<T>, Sacado::Fad::DFad<U> >
+{
+  typedef Sacado::Fad::DFad<typename ProductType<T,U>::type > type;
```
Comment (bangerth) [inline include/deal.II/base/sacado_product_type.h]: And the same here: these typedefs should match what actually happens when multiplying such objects. I think this is what Sacado *should* be doing, but I'm curious if that's what it is *really* doing.

If you want to find out, use a little program like this:
```
template <typename T>
void foo (T);

void bar () {
  DFad<double> dd;
  DFad<float> df;
  foo (dd * df);
}
```
compile, and then use `nm foo.o | c++filt` to see for which type `foo<T>` was instantiated.

File: include/deal.II/base/sacado_product_type.h [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+ * @relates ProductType
+ */
+
+template <typename T, typename U>
+typename ProductType<Sacado::Fad::DFad<T>, typename EnableIfScalar<U>::type>::type
+inline
+operator*(const Sacado::Fad::DFad<T> &left, const U &right)
+{
+  typedef typename ProductType<Sacado::Fad::DFad<T>, U>::type result_type;
+  return static_cast<result_type>(left) * static_cast<result_type>(right);
```
Comment (bangerth) [inline include/deal.II/base/sacado_product_type.h]: ah, I actually see now -- you do implement the product yourself. Then the questions above are moot.

File: tests/base/sacado_product_type_01.cc [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+#include "../tests.h"
+
+template <typename T, typename U, typename CompareType>
+void check()
+{
+  AssertThrow (typeid(typename ProductType<T,U>::type) == typeid(CompareType),
+               ExcInternalError());
+  AssertThrow (typeid(typename ProductType<T,U>::type) == typeid(T() * U()),
+               ExcInternalError());
```
Comment (bangerth) [inline tests/base/sacado_product_type_01.cc]: oh, and clever tests too!

**Pull request merged**